src/eids.js

```diff
@@ -195,3 +196,61 @@ export class EidRewriter {
     return this.ensureUnique(`${eId}_${count}`, false);
   }
 }
+
+export class WorkComponentRewriter {
+  constructor () {
+    this.counters = {};
+    this.skipElements = ['meta', 'preface', 'preamble', 'body', 'mainBody'];
+  }
+
+  /** Rewrites the FRBR URI work components for all attachment nodes in the tree.
+   */
+  rewriteAllAttachmentWorkComponents (element, prefix = '') {
+    this.counters = {};
+    this.rewriteAttachmentWorkComponent(element, prefix);
+  }
+
+  rewriteAttachmentWorkComponent (element, prefix = '') {
+    let tag = element.tagName;
+    if (this.skipElements.includes(tag)) {
+      return;
+    }
+
+    if (tag === 'attachment') {
```

Comment:
You could also do an xpath or getElementsByTagName to get all the `attachment` tags, and only process those, rather than  walking the tree in Javascript.

Reply:
Hmm, I'm wrong: walking the tree is important otherwise we don't have the parent context.